packages/replay-next/src/suspense/HitPointsCache.ts

```diff
@@ -0,0 +1,138 @@
+import { Location, PointRange, TimeStampedPoint } from "@replayio/protocol";
+
+import { MAX_POINTS_FOR_FULL_ANALYSIS } from "protocol/thread/analysis";
+import { createFetchAsyncFromFetchSuspense } from "replay-next/src/utils/suspense";
+import {
+  HitPointStatus,
+  HitPointsAndStatusTuple,
+  ReplayClientInterface,
+} from "shared/client/types";
+import { isThennable } from "shared/proxy/utils";
+import { ProtocolError, isCommandError } from "shared/utils/error";
+
+import { RangeCache, createGenericRangeCache } from "./createGenericRangeCache";
+
+const hitPointCaches = new Map<string, RangeCache<TimeStampedPoint>>();
+
+function getHitPointsCache(location: Location, condition: string | null) {
+  const key = `${location.sourceId}:${location.line}:${location.column}:${condition}`;
+  if (!hitPointCaches.has(key)) {
+    hitPointCaches.set(key, createHitPointsCache(location, condition));
+  }
+  return hitPointCaches.get(key)!;
+}
+
+function createHitPointsCache(location: Location, condition: string | null) {
+  return createGenericRangeCache<TimeStampedPoint>(
+    async (client, range, cacheValues, cacheError) => {
+      const locations = client.getCorrespondingLocations(location).map(location => ({
+        location,
+      }));
+
+      if (condition) {
+        const mapper = `
+          const { point, time } = input;
+          const { frame: frameId } = sendCommand("Pause.getTopFrame");
+
+          const { result: conditionResult } = sendCommand(
+            "Pause.evaluateInFrame",
+            { frameId, expression: ${JSON.stringify(condition)}, useOriginalScopes: true }
+          );
+
+          let result;
+          if (conditionResult.returned) {
+            const { returned } = conditionResult;
+            if ("value" in returned && !returned.value) {
+              result = 0;
+            } else if (!Object.keys(returned).length) {
+              // Undefined.
+              result = 0;
+            } else {
+              result = 1;
+            }
+          } else {
+            result = 1;
+          }
+
+          return [
+            {
+              key: point,
+              value: {
+                match: result,
+                point,
+                time,
+              },
+            },
+          ];
+        `;
+
+        await client.streamAnalysis(
+          {
+            effectful: false,
+            locations,
+            mapper,
+            range,
+          },
+          {
+            onResults: results => {
```

Comment:
I know the analysis manager pre-dates this PR, but I'm confused about the difference between `onResult` (used here) and `onPoints` (used below). Looks like we could almost avoid this fork entirely (and just pass a mapper or an empty string) _except_ for that.

I guess one calls `Analysis.findAnalysisPoints` and the other calls `Analysis.runAnalysis`. Is one more efficient than the other?

Reply:
Yes, finding analysis points is a lot faster than getting the results - the first is just a lookup in the backend while the second needs to run a browser to each point to execute the mapper. So `pointsFinished` will resolve a lot earlier than `resultsFinished`.
When we don't have a `condition` we only need to ask for the points but when we have one we need to execute it for every point to find out if that point should be included.

Reply:
Yep, this is the underlying problem with hit points and print statements to begin with, and this is also why the "breakpoints" implementation that Josh and I had come up with mid-last-year intentionally separated out the "find points" and "run actual analysis" steps in that `protocol/thread/analysis.ts` file.